### ISSUE TITLE
Making typings in fastify.d.ts available under npm @types/fastify

### ISSUE THREAD

**Issue by trivikr** [fastify/fastify#649]
**EDIT**:
The original issue was "About moving fastify type definitions to DefinitelyTyped"
I've renamed it to "Making typings in fastify.d.ts available under npm @types/fastify"
 
#### You have already researched for similiar issues?
Yes.
* The feature request for creating TypeScript definitions was created in #174
* The typings.d file was added in PR #182

#### Are you sure this is an issue with the fastify or are you just looking for some help?

Yes. Posting the question here as it's easier to discuss this in Github issues over [Gitter Chat](https://gitter.im/fastify). Also, it's easier to reference this issue if a PR is created.

#### Is this a security related issue?

No

#### What are you trying to achieve or the steps to reproduce?

As per the [original discussion](https://github.com/fastify/fastify/issues/174#issuecomment-327297748) the type definitions were created in Fastify repo. Because of that, definitions are not available on npm by running command `npm view @types/fastify`

#### What was the result you received?

```
$npm view @types/fastify
npm ERR! code E404
npm ERR! 404 Registry returned 404 for GET on https://registry.npmjs.org/@types%2ffastify
npm ERR! 404 
npm ERR! 404  '@types/fastify' is not in the npm registry.
npm ERR! 404 You should bug the author to publish it (or use the name yourself!)
npm ERR! 404 
npm ERR! 404 Note that you can also install from a
npm ERR! 404 tarball, folder, http url, or git url.
```

#### What did you expect?

I expected type definitions of fastify to be available at `@types/fastify`

#### Context

Question is about location of type definitions, it's not specific to any version

**Comment by jsumners** [comment]
I 100% support moving the types out of core.

**Comment by mcollina** [comment]
What’s the best way to interface with the TypeScript community? We would like to support you folks, but we are getting mixed inputs on where it’s the best spot to place the definitions.

**Comment by trivikr** [comment]
The best spot to place the definitions is in [DefinitelyTyped](https://github.com/DefinitelyTyped/DefinitelyTyped/)

All libraries have their individual folders in DefinitelyTyped, for example:
- [node](https://github.com/DefinitelyTyped/DefinitelyTyped/tree/master/types/node)
- [express](https://github.com/DefinitelyTyped/DefinitelyTyped/tree/master/types/express)
- [hapi](https://github.com/DefinitelyTyped/DefinitelyTyped/tree/master/types/hapi)
- [restify](https://github.com/DefinitelyTyped/DefinitelyTyped/tree/master/types/restify)
- [koa](https://github.com/DefinitelyTyped/DefinitelyTyped/tree/master/types/koa)
- [micro](https://github.com/DefinitelyTyped/DefinitelyTyped/tree/master/types/micro)
- [connect](https://github.com/DefinitelyTyped/DefinitelyTyped/tree/master/types/connect)

I can create folder for fastify in DefinitelyTyped and move the type definitions, if you agree with the decision.

**Comment by dougwilson** [comment]
My understanding is that DefinatelyTyped wants the definitions to be in the projects, and not in that repo. Here is their guide on publishing the types: https://www.typescriptlang.org/docs/handbook/declaration-files/publishing.html

> If you control the npm package you are publishing declarations for, then the first approach is favored. That way, your declarations and JavaScript always travel together.

The project removes ones that include their own typings (see https://github.com/DefinitelyTyped/DefinitelyTyped/issues/13364). There is a lot of discussions in there and the awesome @blakeembrey can explain it better as well.

I have removed various projects from the DefinitelyTyped repo as their typings got included into the main repo, which is their process.

**Comment by trivikr** [comment]
Thanks @dougwilson for the documentation and the issue link.
My question if just about making the fastify typings available in `@types/fastify`, the location of typings file is not important.

For example project babylonjs from DefinitelyTyped/DefinitelyTyped#13364:
* Typings are defined in [babylon.d.ts](https://github.com/BabylonJS/Babylon.js/blob/master/dist/preview%20release/babylon.d.ts)
* The folder [babylonjs](https://github.com/DefinitelyTyped/DefinitelyTyped/tree/master/types/babylonjs) is not present in DefinitelyTyped
* Running `npm view @types/babylonjs` returns details of the typings

I agree with the recommendation by [documentation on publishing types](https://www.typescriptlang.org/docs/handbook/declaration-files/publishing.html) to have types bundled with npm package so that:
* they can be changed when API/types changes in the same package
* developers don't have to edit DefinitelyTyped package separately

How can `@types/fastify` be linked with [fastify.d.ts](https://github.com/fastify/fastify/blob/master/fastify.d.ts)?

**Comment by dougwilson** [comment]
I see the down vote so my post must not have been helpful.

**Comment by trivikr** [comment]
I came across DefinitelyTyped/DefinitelyTyped/pull/15634 in which babylonjs provides it's own types by defining their typingsPackageName in [notNeededPackages.json](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/0ae62b869b05630354cd304be381e02035b99a10/notNeededPackages.json#L129-L134)

As of now, we can add the following code so that fastify typings are available under `@types/fastify` on npm
```json
{
  "libraryName": "fastify",
  "typingsPackageName": "fastify",
  "sourceRepoURL": "https://www.fastify.io/",
  "asOfVersion": "0.27.0"
}
```

What do you think?

**Comment by dougwilson** [comment]
The README in DefinatelyTyped has the following here https://github.com/DefinitelyTyped/DefinitelyTyped#removing-a-package:

> If a package was never on DefinitelyTyped, it does not need to be added to notNeededPackages.json.

Not sure if they would take that change, but doesn't hurt to just do it anyway.

**Comment by nwoltman** [comment]
If you install the latest version `@types/babylonjs` you'll find that the package does not contain any typings. DefinitelyTyped/DefinitelyTyped#15634 was done to remove the typings from the `@types/babylonjs` package so that the only place to get typings in the future would be from the `babylonjs` package.

Also, from the [TypeScript Handbook](http://www.typescriptlang.org/docs/handbook/declaration-files/publishing.html):

> Now that you have authored a declaration file following the steps of this guide, it is time to publish it to npm. There are two main ways you can publish your declaration files to npm:
>
> 1. bundling with your npm package, or
> 2. publishing to the `@types organization` on npm.
>
> If you control the npm package you are publishing declarations for, then the first approach is favored. That way, your declarations and JavaScript always travel together.

So they prefer that declarations are shipped with the package.

**Comment by trivikr** [comment]
Looks like individual packages have to publish their typings to npm under `@types/packagename` if they decide to bundle types with their package
For example, [types-publisher](https://github.com/Microsoft/types-publisher) publishes types from DefinitelyTyped to npm

**Comment by trivikr** [comment]
After going through several resources online on publishing types to npm, the only suggestion is to add a `main` and a `types`/`typings` property to package.json ([Example post](http://blog.bigfont.ca/add-typescript-definition-files-to-the-npm-types-organization/))

fastify already has them added in it's package.json
https://github.com/fastify/fastify/blob/e3b8f8e3d1d547f2569da9d4528d0a904c94011b/package.json#L5-L6

I'm not sure what step we're missing to make these typings available under `@types/fastify`

**Comment by trivikr** [comment]
A [comment](https://github.com/IdentityModel/oidc-client-js/issues/135#issuecomment-269537760) from @andy-ms in one of the discussions

> If a package [bundles](http://www.typescriptlang.org/docs/handbook/declaration-files/publishing.html) its own types, we do not recommend having an @types package too.

Since typings are bundled with fastify, they need not be published under `@types/fastify`
This issue can be resolved

**Comment by trivikr** [comment]
Confirmed with [trivikr/test-fastify-typescript-server](https://github.com/trivikr/test-fastify-typescript-server) that fastify bundled typings work and they're not required to be published under `@types/fastify`

**Comment by mcollina** [comment]
I’m wondering if we need to ship @types/node within the project, or if we can skip it.

**Comment by trivikr** [comment]
As per the recommendation in the [documentation](https://www.typescriptlang.org/docs/handbook/declaration-files/publishing.html), it's advisable to ship type definitions within the project. This way declarations and JavaScript always travel together.

**Comment by jsumners** [comment]
@dougwilson don't mind me. I'm just not a fan of the things. I would have use a ☹️ instead of a 👎, but that isn't possible without a full reply. And a full reply would have been more intrusive to the actual conversation.

**Comment by nwoltman** [comment]
We don't really need to ship within the project, it's just recommended. There are some positive points for moving the typings to `@types/fastify`:

+ We wouldn't need to worry about updating typings when making changes or when releasing a new version of Fastify since that can be done separately (although this could mean TypeScript users will have to wait for updates to the typings after a new release)
+ We'd never need to release a new version of Fastify just to update/fix typings

**Comment by trivikr** [comment]
@jsumners made a good point in related discussion at pinojs/pino/issues/344

> I just really wish my editor would quit going to the TS definitions, when they are included, instead of the actual source code

Adding to it the reasons added by @nwoltman above, and since most popular JavaScript projects maintain their typings in DefinitelyTyped it could be better to move typings to `@types/fastify`

However, the disadvantages of adding type definitions to DefinitelyTyped:
* It's difficult to compare if Typings are in sync. They need to compare the project with it's types in DefinitelyTyped.
* The version numbers of the project and @types are usually not in sync. For example, the latest release of `node` is [v9.4.0](https://github.com/nodejs/node/tree/v9.4.0) while the latest release of `@node/types` is [v9.3.0](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/node/index.d.ts)
* The contributors of the package don't know if their APIs are represented correctly by typings, and it can change any time. For example, I can change competitor package APIs to make it look less attractive to TypeScript developers (rare case, I don't think anyone would do it)

**Comment by jsumners** [comment]
I like the argument made in https://github.com/pinojs/pino/issues/344#issuecomment-357381053

Basically, Fastify is a Javascript project. There is no reason to include non-Javascript within the project. Therefore, I vote the typings be removed.

**Comment by NoNameProvided** [comment]
> Basically, Fastify is a Javascript project. There is no reason to include non-Javascript within the project.

Wow, that's so short-sighted. So basically you say, you don't care about the dev experience of the typescript devs because you use javascript and this repo was written in plain JS. 

As a developer who writes everything in TS, I kept an eye on this discussion. I think with all this listing of pros and cons, this discussion is not about the right topic, you can keep throwing pros and cons into this thread all year. 

The main question is here do you want to make TS developers first-class citizens in the fastify community or not. If you want to include us then play by the rules what the TS community established, or you can roll without supporting TS, but don't do it half-way. @trivikr notes are on the point, typings in under the `@types` namespace are often out of sync, and **bad typings are worse than no typings**, it gives a false sense of security to the TS devs.

**Comment by mcollina** [comment]
@NoNameProvided as you can see, this project includes typings. However it was asked to remove them and move somewhere else by another Typescript user. As I said, I’m
puzzled because it thought we were doing our best to support TS, and yet we are having this conversation. Is there any issues with the current typings? If you think that what we ship are bad typings, feel free to send a PR.

I think we will keep having them here as long as there are people willing to maintain them.

I sincerely do not know where those should live. I think they better live here. However, I would remove the typings for Node.js and pino, and leave them for the end user to install if they want to. Would that be ok for the TS community?

Disclaimer: I’m personally against all forms of transpilation for Node.js applications, and I will likely never use Typescript. I care about all Node.js users, including typescript (and babel, and jsx, and flow, ...) fans.

**Comment by jsumners** [comment]
@NoNameProvided just because the typings are in the core project doesn't mean they will stay synchronized. Regardless of where they are housed, someone has to maintain them. Thus, they can be bad typings here or bad typings there (`@types`).

Why would the types being excluded from this repository exclude TS people? I don't think it would. They are used to getting types from a third party anyway.  What it does do is keep this project focused on its goal.

Further, it's impossible to maintain the types for the object returned by `require('fastify')()`. Look at https://github.com/fastify/fastify-url-data/issues/2. The whole issue is that the plugin decorates Fastify's request object with a new method. By including the plugin in a project, that evidently made it no longer possible to develop against Fastify with TypeScript. That's just one plugin amongst many that have already been written. The problem will only continue to grow.

<details>
<summary>Personal opinion (🔥)</summary>
  
By using TypeScript you create your own set of problems. Chief among them is not having the type declarations available for everything that's ever been written. It is really quite pompous to expect everyone else to conform to your platform choices.
</details>

**Comment by trivikr** [comment]
Hey guys, I think we had enough discussion about where typings should exist.
Here's a general summary about any project:
1. if it's a TypeScript project, no separate typings are necessary.
2. if it's a JavaScript project
  a. if any of the maintainers is a core TypeScript user, typings will be bundled with the project and they'll be kept updated
  b. if all maintainers are JavaScript users, the typings will be rarely touched and it's better to have them moved to DefinitelyTyped 

I've not yet contributed to fastify/pino, and my participation in this project will depend on whether it'll replace express in my personal/work projects.
From the discussion till now, it looks like fastify is falling in 2b. As a TypeScript user, it's better for TS community if a JavaScript project follows 2a

Here are questions for current maintainers @mcollina @jsumners @delvedor @nwoltman :
1. Is it a dealbreaker for you if typings are bundled with your project? Like when it comes to your editor
   * This question is wrt to comment pinojs/pino/issues/344#issuecomment-357223599
    > I just really wish my editor would quit going to the TS definitions, when they are included, instead of the actual source code
2.  If typings get outdated when bundled with the project, can you still keep retain them as some TypeScript user in future is likely to update and send you a PR
    * This question is wrt comment pinojs/pino/issues/344#issuecomment-357327722
    > However, the moment @trivikr and @tniessen stop maintaining them (either officially or they become outdated) and no one else picks up the torch, I would be looking at removing them again
3. If there's a major release (say **N.0.0**), and a TypeScript user finds out that type definitions are not updated and sends a PR, can you make release **N.0.1** just with the typings?

I still think this project will choose 2b, but still wanted to know what maintainers think before finalizing on a decision.

**Comment by jsumners** [comment]
> 1. Is it a dealbreaker for you if typings are bundled with your project? Like when it comes to your editor

Personal opinion: yes. I can't read them and have no use for them. So they only frustrate me.

> 2. If typings get outdated when bundled with the project, can you still keep retain them as some TypeScript user in future is likely to update and send you a PR

I suppose? I really have no idea since I never use them.

> 3. If there's a major release (say N.0.0), and a TypeScript user finds out that type definitions are not updated and sends a PR, can you make release N.0.1 just with the typings?

I imagine that's what would happen, yes.

**Comment by nwoltman** [comment]
1 / 2 - I don't mind as long as someone else maintains the typings (or if someone reminds me to update the typings if I'm about the make a change that will break them).

3 - That's up to whoever has release privileges, but yes that probably would happen.

**Comment by jtlapp** [comment]
FWIW, I landed here to evaluate whether to use Fastify. I made the decision to only develop server-side with static type checking and decided on Typescript. Were the typings maintained by a third party, as they are with HAPI, they would be 2nd-class citizens, as they are with HAPI, and I'd be inclined to find a project that's making assurances about whether it works with Typescript.

I think it's a matter of whether you want to ensure Typescript compatibility.

I see the value in making pure-JS modules because doing so maximizes their reach. Ensuring Typescript compatibility extends the reach further, giving Typescript users the warm and fuzzy they might need to make a commitment to Fastify.

But I agree that you shouldn't have the typings in distro unless you have someone willing to maintain them. They might go in DefinitelyTyped until then, moving to the distro once you have a core maintainer. (But that logic may not apply given that they are already in the distro.)

**Comment by jsumners** [comment]
We have someone responsible for the types, @evanshortiss. But they still get out-of-sync. The majority of regular contributors do not use Typescript. Thus the types usually don't get updated until someone comes along and points out an issue with them. At which point Evan is pinged and some things happen.

It is my opinion that *the current situation is equivalent* to the types being in the DefinitelyTyped repository instead of here.

It isn't that the types are a "2nd-class citizen", it's that this is not a Typescript project.

Answer me this: if the types didn't exist at all, could you still use Fastify in a Typescript project? If so, is the only reason they're used is to get pretty popups for method signatures? That being the case, would jsdoc blocks be equivalent?

**Comment by jtlapp** [comment]
> But they still get out-of-sync... It isn't that the types are a "2nd-class citizen", it's that this is not a Typescript project.

Gotcha. The reason I'm looking for 1st-class support is exactly so that they are in sync. Including them in the distro doesn't by itself solve that problem.

> if the types didn't exist at all, could you still use Fastify in a Typescript project?

If people want Typescript to tell them when they've got a type wrong, they have to include the `--noImplicitAny` compiler option. Otherwise failures to type quietly pass as properly typed. When this option is set, every import of a package without typings becomes an implicit `any` type and won't build.

The workaround is simple: cast the import as type `any`, but then no static type checking occurs for interactions with the framework. That's not ideal, but it may be a reasonable tradeoff if it's important enough to be using Fastify.

(Actually, devs will likely have to sprinkle the `any` typecast throughout the code, where it interfaces Fastify. It's anathema to why we use Typescript, but might be a necessary pill to swallow.)

**Comment by jsumners** [comment]
Was that a "yes" or a "no"?

**Comment by jtlapp** [comment]
> if the types didn't exist at all, could you still use Fastify in a Typescript project?

Yes.

> If so, is the only reason they're used is to get pretty popups for method signatures?

No. 

> That being the case, would jsdoc blocks be equivalent?

No.

**Comment by jtlapp** [comment]
This is an aside, but to help you understand why people use Typescript, I'll explain.

> if the types didn't exist at all, could you still use Fastify in a Typescript project?

Many of us who use Typescript come from a background in Java, C++, or other strongly-typed language. There are many classes of errors that a compiler will find and report before the program ever runs that you only encounter at runtime (or in tests) in plain Javascript. We find that really annoying and a huge waste of time, for these kinds of errors.

Examples of errors that the compiler will catch include assigning a number to a variable that should only ever hold a string, referencing an object property that does not exist (misspellings!), accessing a property expecting the value to be assigned when it hasn't yet been, accessing a property as if it were a string or a number when it's really just null. Save a lot of headache to fix these before the first run.

I consider all that of minor benefit compared to the benefit I get when I need to refactor. On my first pass of coding, it's easy for me to keep everything in my head and all my code consistent. But later when I (or someone else) needs to make a change that potentially affects many places of code, I'll do the greps and seds and search-and-replaces to make the necessary changes, but invariable I miss a few things. The Typescript compiler will often tell me what I missed before I try to run the refactored code.

BTW, I have pop-ups turned off. I find them annoying and in the way of seeing the code.

**Comment by jsumners** [comment]
Okay, I'm calling for a vote. As of this writing https://github.com/fastify/fastify/pull/635 has been open for a week without any response, and this discussion merely looping around.

@fastify/fastify and community, please cast your vote on this comment by 2018-01-17 14:00 UTC.

👍 => remove the types from this repository and move them to the `@types` repository
👎 => keep the types in this repository

**Comment by jtlapp** [comment]
My take: keeping typings in fastify proper gives Typescript devs like me the warm-and-fuzzy that there is a serious commitment to keeping typings in sync, that it is safe to in turn commit to fastify.

My experience is that 3rd parties often quietly stop maintaining typings.

**Comment by delvedor** [comment]
I don't have a strong opinion on this. I'm not a Typescript user.
If have the typings in this repo will keep typescript devs *"warm-and-fuzzy"* it's ok for me keeping the types here. But it must be clear that **we do not guarantee** that the types will always be in sync. If someone wants to send a pr to update them we will happily accept it and release a patch update.

**Comment by StarpTech** [comment]
It's easier to move them out of fastify so we don't have to deal with them. Any contributor can create a PR in @types repo.

https://github.com/DefinitelyTyped/DefinitelyTyped#make-a-pull-request

**Comment by mcollina** [comment]
I prefer if they are left in here, as they will create a warm and fuzzy environment for typescript users.

However I would remove @types/node and @types/pino. That simplify maintenance for us.

**Comment by nwoltman** [comment]
I'm going to abstain from voting. I don't have a strong opinion in either direction.

**Comment by jtlapp** [comment]
Another point to consider, from the [DefinitelyTyped FAQ](https://github.com/DefinitelyTyped/DefinitelyTyped#faq):

> I've submitted a pull request. How long until it is merged?
>
> It depends, but most pull requests will be merged within a week.

At least there's a chance for in-repo typings to stay in sync with changes.

**Comment by jsumners** [comment]
At this point, please just cast a vote on https://github.com/fastify/fastify/issues/649#issuecomment-357687748.

**Comment by trivikr** [comment]
@jtlapp About pull requests getting merged in DefinitelyTyped as per [FAQ](https://github.com/DefinitelyTyped/DefinitelyTyped#faq)

> I've submitted a pull request. How long until it is merged?
It depends, but most pull requests will be merged within a week.

This depends on how fast the type definition maintainers review the incoming PRs.
* Example PR DefinitelyTyped/DefinitelyTyped/pull/22865 was reviewed by type definition maintainers and merged on the same day
* Example PR DefinitelyTyped/DefinitelyTyped/pull/22766 wasn't reviewed by type definition maintainers for 5 days, so DefinitelyTyped maintainer was involved.

The details of type definition maintainers are present in the documentation.
For example, for express it's borisyankov ([code comment](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/29c6e0f81ab314cd1d1f17ab92f59bb87f6a5c3f/types/express/index.d.ts#L3))

**Comment by StarpTech** [comment]
Allow me to summarize:
- No one in the Fastify Team works with Typescript.
- We don't want to update the definitions.

> I see the value in making pure-JS modules because doing so maximizes their reach. Ensuring Typescript compatibility extends the reach further, giving Typescript users the warm and fuzzy they might need to make a commitment to Fastify.

We develop Fastify in Javascript because it's idiomatic and minimal. Typescript is a superset of Javascript you don't need a definition file for everything it is the pressure of a typescript user.

**Conclusion:**
IMHO we should maintain the definition files until 1.0 was released and after that, we should delegate it to the community. The best way is to manage them via http://definitelytyped.org/ and each typescript user can create a PR.

**Comment by jtlapp** [comment]
@delvedor wrote:

> But it must be clear that we do not guarantee that the types will always be in sync.

I think keeping the typings in the repo is a decision to make an effort to reach out to Typescript people to keep typings in sync with the releases.

I think that's the question: Are the maintainers willing to make an effort to keep the typings in sync?

If the answer is no, the typings do not belong in the repo. It creates the wrong expectations, sends the wrong message about the level of support provided for Typescript.

(I'm not voting because I'm evaluating Fastify, not yet a member of this community.)

**Comment by mcollina** [comment]
@evanshortiss can you please chime in?

**Comment by mcollina** [comment]
> Are the maintainers willing to make an effort to keep the typings in sync?

This is the wrong question. As the asnwer is no, because the maintainers are not willing to make an effort on maintaining anything. Software is provided as-is, as the license states.

What is emerging from this discussion is that there are a significant number of people in the TS community that would like to have the typings, but they would not like to spend the time maintaining them. I am a bit puzzled, and this is not how OSS work.

The rules for contributing to this project are VERY LAX, and we give to anybody who is active the commit bit. 

@evanshortiss is a full member of this community, and he helped so far in the maintenance of these. If @evanshortiss is not active enough to maintain these, why don’t somebody else who cares (and there seems to be people using thise typinings) willing to step up and maintain them?
In any case, we should wait for him to join this conversation, so I propose to extend the deadline accordingly.

I don’t think that maintaining them externally would help at all in providing a good experience to the end user. I think having first-class Typescript support should be a feature of this framework, if there is someone in the TS willing to step up. As of today, the typings are in sync and somehow well maintained, and I still do not get why TS users complained in several occasions.

**Comment by jtlapp** [comment]
The way it works in other repos where the maintainers are not adept at Typescript is that they reach out to known contributors who are adept to get a check on whether things are up-to-date.

That's all "making an effort" needs to mean.

What "making an effort" precludes is publishing a new version of the Fastify API without checking with anyone to see if the Typings need to be updated.

"Making an effort" doesn't mean "making a guarantee." No one can make a guarantee in software for anything, anyway.

It'll be a month or two before I'm certain that I'm on board with Fastify, so I myself couldn't commit to maintaining typings for a while.

**Comment by mcollina** [comment]
> The way it works in other repos where the maintainers are not adept at Typescript is that they reach out to known contributors who are adept to get a check on whether things are up-to-date.

Why they should be pinged and should reach out? Can TS people watch the repo and review PRs and help as everybody else? I think no one will complain if someone add a note saying how to fix the types. Or send a PR themselves?

I am not blocking a release or a merge waiting for someone to check if the types are ok. Bugs happens all the time, and this is no different. Why would you be entitled to anything more?

**Comment by jtlapp** [comment]
> Can TS people watch the repo and review PRs and help as everybody else?

Yes of course. If you have someone keeping tabs on things, you won't need to reach out.

> Why would you be entitled to anything more?

When someone knows that something breaks, they tend to report it to the person who can fix it. Changing the public API breaks the typings. As with anything else, report it to whoever can fix it, if that person isn't already on top of it.

I'm feeling pretty queasy about now.

**Comment by NoNameProvided** [comment]
> What is emerging from this discussion is that there are a significant number of people in the TS community that would like to have the typings, but they would not like to spend the time maintaining them. I am a bit puzzled, and this is not how OSS work.

That's not true (at least from my side), as it's visible from this conversation. I do care and as I Typescript user I am willing to "do my best" to keep typings up-to-date. Thats why I opened #666 for example.

_opinion with personal feelings:_

I really appreciate the work the fastify team does, it's a fascinating piece of software, and I am really looking forward to using it and after some experimenting hopefully putting it under the hood of [typestack/routing-controllers](https://github.com/typestack/routing-controllers). On the other side, I have this conversation, where the majority of the fastify team claims they don't know Typescript, yet don't want to listen to Typescript users about how typing should be provided for this library.

**Comment by mcollina** [comment]
It seems to me the exact opposite. It seems we would like to provide typescript support, and there are very few people willing to help and they just complain about it. Currently there are no open issues about the typings having bugs, so what is the problem?

**Comment by NoNameProvided** [comment]
The vote on moving it to `@types/fastify`, that is a one-way road, once you moved it you can never move it back without causing a lot of confusion. (Because the package cannot be removed from npm.)

**Comment by StarpTech** [comment]
@NoNameProvided everybody can create a PR in DefinitelyTyped and it will be published. We should go one way but document it.

**Comment by delvedor** [comment]
I've opened  a pr with @fastify/core's proposal.
https://github.com/fastify/fastify/pull/669

**Comment by jsumners** [comment]
Given the above mentioned PR, I am closing the poll. At this time it is tied 3 to 3.

I will also say that you don't have "to be a member of the community" to submit fixes via a PR. There are plenty of times that I submit a fix to a project while evaluating it and never even go back to it.

**Comment by jtlapp** [comment]
I'll start porting my server to Fastify later this week and will offer typings PRs for anything I notice that's out of sync. I'm very much looking forward to it, as it means peeking at the internals.

**Comment by evanshortiss** [comment]
> What is emerging from this discussion is that there are a significant number of people in the TS community that would like to have the typings, but they would not like to spend the time maintaining them. I am a bit puzzled, and this is not how OSS work.

Unfortunately I find myself agreeing with this. Multiple TS users seem to have found issues but not offered solutions; **some** being key here, others have been helpful. We can only speculate as to the reason for this, but it's a problem.

> The vote on moving it to @types/fastify, that is a one-way road, once you moved it you can never move it back without causing a lot of confusion. (Because the package cannot be removed from npm.)

I believe deprecate is an option to resolve this. `@types/commander` has done so and you'll receive warnings as a result.

> Further, it's impossible to maintain the types for the object returned by require('fastify')(). Look at fastify/fastify-url-data#2. The whole issue is that the plugin decorates Fastify's request object with a new method. 

[This PR](https://github.com/fastify/fastify-url-data/pull/4) demonstrates otherwise.

@trivikr, I don't understand why this issue is necessary, and it seems like you have been contradicting yourself? The Issue states:

> I expected type definitions of fastify to be available at @types/fastify

But then you stated:

> I agree with the recommendation by documentation on publishing types to have types bundled with npm package

Then you added this:

> As of now, we can add the following code so that fastify typings are available under @types/fastify on npm

This actually causes a deprecation warning and we should not do this since it would be inaccurate. Here's the output for `npm i @types/babylon`

```
npm WARN deprecated @types/babylonjs@2.4.1: This is a stub types definition for BabylonJS (http://www.babylonjs.com/). BabylonJS provides its own type definitions, so you don't need @types/babylonjs installed!
```

@trivikr can you clarify what the actual issue is since I'm not sure I follow as per the points above?

**Comment by evanshortiss** [comment]
As to the vote for decoupling definitions, I will vote in favour of whatever is easiest for the core maintainers here i.e if the majority of core say remove then I would not object.

TypeScript users will be able to develop just fine if the types are bundled with fastify, or separately, it's just convenient to bundle them but not if it is causing pain for the primary maintainers.

**Comment by jtlapp** [comment]
@evanshortiss, do you know how we'd arrange to have the typings publish concurrently with publications of new Fastify releases? Would it just need an owner who approves (and hopefully provides!) typings PRs in a timely fashion, so that we're not waiting on DefinitivelyTyped folks to approve and merge the PRs -- which I don't see how they'd be qualified to do that anyway.

If both scenarios require someone committing to keeping the typings concurrent, and there is such a person, I wonder if life would be more pleasant moving them out. This group is not shy about showing resentment for their being here or for having to bend in anyway for Typescript. (Not just this thread -- I've racked up a wondrous thumbs-down collection!)

If it would relieve tensions to move them out, maybe that should trump any other reason for keeping them. It's going to be hard to keep a Typescript person around always being on the defensive.

**Comment by trivikr** [comment]
@evanshortiss Actually, my knowledge about typings has evolved over this issue as follows:
* the issue was originally cut as I was not able to find typings for `@types/fastify` as compared to other competitors like express, koa, hapi etc (as explained in [comment](https://github.com/fastify/fastify/issues/649#issuecomment-357115864))
* then I realized it was my mistake. The bundled typings also work, and is actually recommended by TypeScript as explained in [comment](https://github.com/fastify/fastify/issues/649#issuecomment-357158493) after which issue was closed
* then there were some follow up discussions and it appeared that it's easier for current fastify maintainers if type definitions are moved out ([comment](https://github.com/fastify/fastify/issues/649#issuecomment-357453369)) so I raised three questions just to be sure
* after that the discussion became serious about what TS users want vs how JS maintainers should behave leading to [this vote](https://github.com/fastify/fastify/issues/649#issuecomment-357687748). I personally voted for moving out the type definitions so that it's easier for core maintainers to work on their project
* In conclusion @fastify/core decided to continue retaining type definitions in #669 

I apologize for creating confusion. It wouldn't have happened if I'd gone through DefinitelyTyped documentation or [TypeScript.md](https://github.com/fastify/fastify/blob/master/docs/TypeScript.md) before cutting this issue.

This issue can be closed.
I'm eagerly waiting v1.0.0 release to try out fastify in my projects!

**Comment by evanshortiss** [comment]
@trivikr thank you for the clarification; I think the large number of comments made it hard to follow the thought process, but this makes sense now 😄 

> If it would relieve tensions to move them out, maybe that should trump any other reason for keeping them. It's going to be hard to keep a Typescript person around always being on the defensive.

@jtlapp this is more or less how I feel. And reasons highlighted by here by @nwoltman are solid arguments among other things. If we were talking about Babel etc. I'd probably share @jsumners frustrations since, unlike TS, I think they don't add value. This is also something that can be revisited as the TS ecosystem matures.

> do you know how we'd arrange to have the typings publish concurrently with publications of new Fastify releases

Not 100% sure really. I guess we'd need a best effort approach where if the MINOR or MAJOR version is being bumped then a collaborator who's familiar with TypeScript can review changes and update the types accordingly?

**Comment by jtlapp** [comment]
@evanshortiss wrote:

>Not 100% sure really. I guess we'd need a best effort approach where if the MINOR or MAJOR version is being bumped then a collaborator who's familiar with TypeScript can review changes and update the types accordingly?

We'd need pre-release tagging too, me thinks, such as `1.1.0-alpha`, `1.1.0-beta`, maybe `1.1.0-rc`. Some clue that the API has stabilized but the release has not yet been made.

**Comment by delvedor** [comment]
Closed by https://github.com/fastify/fastify/pull/669.
